gnd/src/main/java/com/google/android/gnd/repository/local/RoomDataStore.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gnd.repository.local;
+
+import androidx.room.Room;
+import androidx.room.Transaction;
+import com.google.android.gnd.GndApplication;
+import com.google.android.gnd.repository.local.Edit.Type;
+import com.google.android.gnd.vo.Feature;
+import com.google.android.gnd.vo.Point;
+import io.reactivex.Single;
+import javax.inject.Inject;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/** Implementation of local data store using Room to map objects to relational MySQL database. */
+public class RoomDataStore implements LocalDataStore {
+  private static final String DB_NAME = "gnd-db";
+  private static final String LOCATION_JSON_KEY = "ll";
+
+  private final GndRoomDatabase db;
+
+  @Inject
+  public RoomDataStore(GndApplication app) {
+    this.db =
+        Room.databaseBuilder(app.getApplicationContext(), GndRoomDatabase.class, DB_NAME).build();
+  }
+
+  @Override
+  @Transaction
+  public Single<Feature> addNewFeature(Feature feature) {
+    return db.featureDao()
+        .insert(toFeatureEntity(feature))
+        .map(newId -> feature.toBuilder().setLocalId(newId).build())
+        .flatMap(f -> db.featureEditDao().insert(toFeatureEditEntity(f)).map(__ -> f));
```

Comment:
Let's avoid double underscore for unused variable names, as per https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names (and because there are rumblings of single underscore becoming a special character in future Java versions, and it seems best to avoid something that will be visually almost the same).

Reply:
It appears Java banned them because they're on a path to using them specifically for this purpose: http://openjdk.java.net/jeps/302

Once added to the language, it would be a natural and easy change from `__` to `_`. This is also probably why many online examples of RxJava and other functional APIs use this syntax.

Wdyt?

Reply:
Just double checked, and seems Google architecture examples also rely on this convention, for ex.:
https://github.com/googlesamples/android-architecture/blob/dev-todo-mvvm-rxjava/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.java

Reply:
Just chiming in on this one since it applies to general practices throughout the codebase. `_` as a non-bindable name/"don't care" pattern match is a widely adopted convention in FP languages and libraries. Typically I'd agree with Navin and drop `__` to avoid ambiguity once `_` is adopted into Java, but, since we're talking about lambdas, the scope of `__`'s will be very limited anyway, plus we're not actually using the binding, so even if we were to miss a couple of conversions to `_` when it hits it wouldn't be a huge deal. 

I'm for adopting `__`. To me it's a clear signal that the binding isn't going to be used in the lambda body. It's better than using an actual name, e.g. ('x' or 'notused' or whatever) which would for often lead me to wonder if someone forgot to use the binding in the function body. 

That said, I think Java's method reference form, `(something::function)`, should be preferred where possible. That's not possible, of course, for cases in which we do more than call a single function in the lambda.

For the particular code we're talking about here, I'd wonder whether there's a better idiom than a lambda since you're (I think? just skimmed, so I might be wrong) making direct replacements with map. For example, in something like pseudo Haskell you could replace the lambda here with `map((const f))` Const is just `const:: a -> b -> a` but this particular use of it relies on higher order functions and currying.

Guava Functions does have [constant](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Functions.html#constant(E)) function so something like `.map(Guava.Functions.constant(f))` should work? And eliminate the lambda in this case. No guarantees. I'm still not super familiar with Java's approach to FP and a lot of intuitive solutions in FP style languages are stymied in Java from time to time.

All that said, I have no qualms with the lambda since it's plenty clear to me.

Reply:
I initially used `__` because in addition to being a common pattern found throughout Google and RxJava code samples, and similar to `_` in other languages, the convention makes it explicit that the lambda param is intentionally being ignored. If there's a better identifier that makes this fact clear I'd be happy to replace it globally.

Re this specific case, I want to condition the result of the `Single<Feature>` on the outcome of the inner `insert()` call. There's almost certainly a better way of doing this in Rx. Looking into it now.

Reply:
Totally—I might’ve been unclear with my comment, I’m onboard with __ and I agree that it makes the lack of usage of the binding explicit. 

Back to the single, I’m dealing with a similar thing right now for the leaky subscriptions refactor—finding a way to conditionally manipulate flowables and the like. 

Thus far switchIfEmpty, concat, merge, ambArray, and some others seem potentially useful depending on the situation. 

Curious to see what you find out!

Reply:
See latest commit; this was the closest I could find to using pure Rx semantics to achieve the same thing.